### DISCUSSION TITLE
Map conditionals

### DISCUSSION THREAD

**Discussion by miguelalbuquerque** [scandum/tintin#53] [Ideas]
Hi there,
First: amazing client. Thank you.

In the MUD I play different variables can condition what map areas I can access, which exists I can take, etc. These can be my current alignment towards a certain faction, my race (guards would stop me or not), etc.

Q: Is there any way we can establish a room exit to have it **evaluate a condition** to "allow it" or not? (a single condition, which you can in-line or call a function or whatever, same syntax as #IF {condition}).

I thought about programming a little script to "goto" a certain number of hardcoded vnums, and sets the room or exit flags to avoid. But that changes the actual map in memory. I have it set to auto-save on session logout, as I keep on adding stuff, and this seems like a non-efficient approach.

KBTIN, for instance, has condition based on its graph traversing aliases feature, but I really love the mapper in TINTIN++.

Thanks!

**Comment by scandum** [comment]
You can set data on rooms and exits, next code a loop to run through each room/exit to set the avoid flag.

Commands of interest:
```
#map list {variable} {varname}
#map set roomdata {simple or nested variable}
#map set roomdata {varname}
#map exit <dir> set {simple or nested variable}
#map exit <dir> get {varname}
```
Also an option to store the data in a table. Let me know if you need a more practical example.

**Reply by miguelalbuquerque** [reply]
Thanks for the answer!
This will help, if I have to set/unset avoid flags in bulk.

What I was asking, however, is how to have the avoid flag being based on a dynamic condition, not a binary hard-set yes or no.
The main idea is to avoid changing the map, as doing so will basically mean I have constantly a file.map with a bunch of edits every time I log with a new character (I keep it as a git repo).
I hope I was clear.
I that's not possible, do consider it an idea request :)

Thanks

**Reply by scandum** [reply]
The idea would be that you'd have a 20-40 line alias to dynamically update the avoid flags once per character login, with the map file containing the necessary meta data.

I agree that it could be a useful feature, though tunneling anything through the scripting engine would be a definite performance bottle neck, especially combined with the brute force nature of path finding.

I'll keep the idea in mind of course, it's been proposed before, but to actually implement it would likely be rather cumbersome as well.

**Comment by rigrig** [comment]
How about `#map list {avoid...} {...}`?

I'm trying to generate "meaningful" directions, something like "The palace is 90 rooms south of here".
My problem is that some cities are ~100 rooms across, but the surrounding "terrains" area has larger-scaled rooms. i.e. the shortest path between the north and the south part of town is often `leave by the north gate, walk around the city, enter through the south gate`, something like `n;n;n;se;sw;n;n;n`. "Adding up" those directions results in `4 north`.
So I get "The palace is 4 rooms north of here", which is rather unhelpful when trying to get an idea of where in the city I am.

I *do* want take the shortcut when walking somewhere, so setting all city->terrains exits to `avoid` is not an option. (And I've mapped ~40000 terrains rooms (and there are many more), so looping through and temporarily flagging them all as `avoid` takes too long.)

It would be nice if I could do `#map list {avoidroomarea} {terrains}`, and it would avoid finding any paths through rooms with `roomarea = terrains`.

And instead of running a script for every room, @miguelalbuquerque could maybe store stuff like factions/alignment in `roomnote`s, and then do something like `#map list {avoidroomnote} {%*enemy_faction%*}` ?

**Reply by miguelalbuquerque** [reply]
So far:
I've implemented a mechanism to store a nested (map) roomdata in rooms, to denote that limitations/conditions they have in terms of mud specific checks.
But yes, then I iterate through all rooms to check that logic against my current character state. This has 2 problems: it changes the map avoid and weight flags if I re-save it. But also, given the map has 100k rooms, it's relatively fast in my main computer, but impossibly slow on my mobile device (emulated shell).

Having a mechanism in the route calculation to use current state can be an added value, and would only imply a performance penalty if actually called (depending on implementation: i.e., a weight being a function instead of a value would only need to funnel through the scripting engine that one room/exit or routes going through that one room/exit, which I'm obviously imagining is an oversimplification with current logic).
To put it in perspective that particular map has about 20 rooms out of 100k with that logic. Chokepoints (bridges, city gates, etc.)

**Comment by scandum** [comment]
@rigrig I assume you're using `#map find <room>` followed by `#path describe` ? It does inform you about the length of the path.

Allowing a terrain based weight could be useful. I'll add it to my todo list. Performance is an issue when it comes to pathfinding, so I'm hesitant about adding a scripted solution when lower level solutions might suffice.

**Reply by rigrig** [reply]
Well, I'm using `#map list` and converting `dist` and `x,y,z` myself, but it gives similar results as `#path describe`.
I think the problem with adding a weight is that it would mess with regular pathfinding. Unless you meant adding it as parameter to the `#map list`/`#map find` commands?


For example: [my script](https://git.tubul.net/tt_dw/scripts/src/branch/trunk/scripts/map/describe.tin#L38) shows this just inside the (northern) Deosil gate: (magenta marker on [this map](http://dw.daftjunk.com/show.php?map=Ankh-Morpork&points=748,46,18,729))
> You are in Ankh-Morpork, 35 rooms south of the Mended Drum;  Jonas' vault is southeast of here, **Sam Slager's sausages** is **north** of here, the temple of Fish is north of here, the Deosil Gate is directly north of here, the terrains lie two rooms to the north, the Least Gate is one room southeast of here and the intersection of Endless Street and Pallant Street, at the Least Gate is two rooms to the south.

The path to "Sam Slager's Sausages"([blue marker](http://dw.daftjunk.com/show.php?map=Ankh-Morpork&points=748,46,18,729)) is `n;n;nw;journey w;s;e;journey w;w;sw;sw;journey se;ne;e;e;nw;n;n;nw;n;n;n;nw;ne;ne;nw;s`, and `#path describe` gives 
> The path is 26 rooms long and the destination lies 11 rooms to the north from you at the same level.

If I speedwalk there, that *is* the path I want to take.
But in that description, I want it to say "Sam Slager's sausages is southwest of here". (Actually, I don't want it to show up at all, because other interesting rooms are closer-by if you avoid leaving the city)

**Reply by scandum** [reply]
I wonder if I could add a weightless mode that treats each room, include void rooms, as having a weight of 1.

Assuming you've extended the path around the city with void rooms, that should provide a purely spatial #map list/find.

Nice work on the describe feature, I've had it on my todo list for a while.

**Reply by rigrig** [reply]
(Whoops, I thought I replied earlier, but apparently I didn't press the button)

> Assuming you've extended the path around the city with void rooms

I'm afraid it isn't that simple: the terrains is basically a grid of rooms with (mostly) 8 exits each (I haven't properly linked them all yet):
![image](https://github.com/scandum/tintin/assets/2941436/dffb4a19-c201-45ee-962d-234b2c4bc905)
That hole in the middle contains the city, with all the gate exits flagged as `hidden`. To get the scaling right, I'd have to insert ~6 void rooms everywhere, which didn't seem like a great idea.

And to complicate things: the terrains have a `journey` feature, which can take you up to 11 rooms, either in a straight line, or following the road (red). I've only added a few of those to my map yet, but it works from any room...

I understand why you don't want to complicate the pathfinding though, and realized that I could solve my problem by making a list of "city gate" *exits*, and temporarily toggling those to `avoid` (instead of all the terrain rooms).


> Nice work on the describe feature, I've had it on my todo list for a while.

Thanks, I got the idea when I saw `#map landmark` :wink:
